src/main/java/uk/gov/register/core/RegisterContext.java

```diff
@@ -62,7 +61,7 @@ public Flyway getFlyway() {
         return flyway;
     }
 
-    public Register buildOnDemandRegister() {
+    public synchronized Register buildOnDemandRegister() {
```

Comment:
I'm concerned about `synchronized` here, and I think we might be able achieve what we want using something like `AtomicReference`s instead. In any case, where we have concurrency concerns (which we certainly do here), we should think about how we document those concerns so that it's clear why the various concurrency primitives are present and how to avoid introducing subtle bugs when maintaining this code.